SIL.Media/Naudio/AudioRecorder.cs

```diff
@@ -279,44 +326,82 @@ public virtual void BeginRecording(string waveFileName)
 			BeginRecording(waveFileName, false);
 		}
 
+		private class InvalidRecordingStateException : InvalidOperationException
+		{
+		}
+
 		/// ------------------------------------------------------------------------------------
 		public virtual void BeginRecording(string waveFileName, bool appendToFile)
 		{
-			lock (this)
+			try
 			{
-				if (_recordingState == RecordingState.NotYetStarted)
-					BeginMonitoring();
+				// This is intentionally done outside of the lock obtained below. In the
+				// unusual case where two different threads (one perhaps being the UI thread)
+				// are trying to start recording (or one is trying to start monitoring
+				// while another tries to start recording). If they are both trying to start
+				// recording, we will end up throwing an InvalidOperationException on one of
+				// them, but we don't care which one. If one is merely trying to start
+				// monitoring and it happens to get there first, the other will find it is no
+				// longer in a valid state to begin monitoring but will happily find that it
+				// is now in a valid state to begin recording. If the one that is trying to
+				// begin recording gets there first, the other one will get the exception.
+				// In any case, handling an InvalidOperationException is the caller's
+				// responsibility.
+				BeginMonitoringIfNeeded();
```

Comment:
The comment above is very helpful. But it still worried me that this code is ignoring a return result that is generally used to report a problem. What you've done is valid...the false return does not necessarily mean a problem, just that monitoring was not begun. But it might be worth adding, "The return result is intentionally ignored. It doesn't matter whether BeginMonitoringIfNeeded started monitoring, found we were already monitoring, or found us in some other state. What matters is the test (below, inside the lock) that we end up in the right state to start recording."